Refactor llcp pdu length validation

erbr-ot opened the pull request:
Using a macro instead of repeated code for length validation

erbr-ot commented:
sits on-top of SCA PR, will rebase once merged

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -1178,290 +1210,192 @@ static bool pdu_is_terminate(struct pdu_data *pdu)
 	return pdu->llctrl.opcode == PDU_DATA_LLCTRL_TYPE_TERMINATE_IND;
 }
 
+#define VALIDATE_PDU_LEN(pdu, type) (pdu->len == sizeof(pdu->llctrl.type) + 1U)
```
```suggestion
#define VALIDATE_PDU_LEN(pdu, type) (pdu->len == PDU_DATA_LLCTRL_LEN(type))
```

erbr-ot replied:
Ack

Thalley commented:
> sits on-top of SCA PR, will rebase once merged

Can you please provide at link to the other PR? Also suggest to mark this as draft until the other PR has been merged

hermabe commented:
> > sits on-top of SCA PR, will rebase once merged
> 
> Can you please provide at link to the other PR? Also suggest to mark this as draft until the other PR has been merged

It is also possible to set the target branch of this PR to the branch of the other PR. Then the diff will show up correctly and once the other PR is merged the target branch of this PR will be changed to main.

thoh-ot reviewed (changes requested):
Change is fine for non-test code, but for tests/bluetooth/controller/common/src/helper_pdu.c I'm a bit "worried" that this change uses `PDU_DATA_LLCTRL_LEN` for both the `helper_pdu_encode_XXX` functions and the `helper_pdu_verify_XXX` functions.

If an error was to be introduced in `PDU_DATA_LLCTRL_LEN`, these tests would no longer catch the length errors - as it's a fairly simple macro the chances are pretty low - I'm thinking in broader terms.

In my world test code is close to the metal and sometimes at bit cumbersome due to the fact that it cannot rely on the constructs it's trying to test and verify - otherwise there's no point in testing if both ends of the test has the same error and therefore self corrects.

Would like to hear other opinions.

Thalley commented:
> Change is fine for non-test code, but for tests/bluetooth/controller/common/src/helper_pdu.c I'm a bit "worried" that this change uses `PDU_DATA_LLCTRL_LEN` for both the `helper_pdu_encode_XXX` functions and the `helper_pdu_verify_XXX` functions.
> 
> If an error was to be introduced in `PDU_DATA_LLCTRL_LEN`, these tests would no longer catch the length errors - as it's a fairly simple macro the chances are pretty low - I'm thinking in broader terms.
> 
> In my world test code is close to the metal and sometimes at bit cumbersome due to the fact that it cannot rely on the constructs it's trying to test and verify - otherwise there's no point in testing if both ends of the test has the same error and therefore self corrects.
> 
> Would like to hear other opinions.

If I am understanding you correct, you are arguing that using the same function/macro/whatever to calculate the value and to verify that the calculation is correct, may not catch any issues. I can fully agree to that, basically you don't want to have a test like

```c
int foo()
{
    return val;
}

int test_foo()
{
    foo_val = foo();
    zassert_equal(foo_val, foo());
}
```

As any errors introduced in `foo` would not be caught.

erbr-ot commented:
I agree with @thoh-ot and @Thalley  - updated PR to keep the test code as is.

cvinayak commented:
Added DNM as there is only one squashed commit in this PR. Please rebase.

thoh-ot commented:
> If I am understanding you correct, you are arguing that using the same function/macro/whatever to calculate the value and to verify that the calculation is correct, may not catch any issues. I can fully agree to that, basically you don't want to have a test like...

@Thalley That was exactly what i was trying to say.

erbr-ot committed:
Bluetooth: controller: rename pdu_data_llctr struct SCA member

Rename struct member to adhere to general naming pattern

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
Bluetooth: controller: refactoring pdu length validation

Using a macro instead of repeated identical code for validation
Use macro for pdu length field encoding

Signed-off-by: Erik Brockhoff <erbr@oticon.com>
